Remember me strategy

rgraff opened the pull request:
This is a RFC as I continue to implement a Remember Me strategy.


  ```elixir
  defmodule MyApp.Accounts.User do
    use Ash.Resource,
      extensions: [AshAuthentication],
      domain: MyApp.Accounts

    attributes do
      uuid_primary_key :id
      attribute :email, :ci_string, allow_nil?: false
    end

    authentication do
      strategies do
        password do
          identity_field :email
          hashed_password_field :hashed_password
          hash_provider AshAuthentication.BcryptProvider
          confirmation_required? true
        end

        remember_me do
          cookie_name :remember_me
          cookie_options [
            max_age: 30 * 24 * 60 * 60, # 30 days
            http_only: true,
            secure: true,
            same_site: :lax
          ]
          remember_me_field :remember_me
          token_lifetime {30, :days}
        end
      end
    end
  end
  ```

zachdaniel commented on lib/ash_authentication/strategies/remember_me/dsl.ex:
```diff
@@ -0,0 +1,52 @@
+          doc: "The name to use for the cookie. Defaults to `remember_me`",
+          default: :remember_me
+        ],
+        cookie_options: [
```
I'm wondering if we should just take a cookie name here and have them set up the cookie in their endpoint?

rgraff replied:
@zachdaniel nevermind. I have an idea and will propose.

rgraff commented:
Video of the developer experience and demo of functionality.
https://www.loom.com/share/1e8882db4a274012bc3c2c84b2966dac?sid=2768ef79-b552-4342-991d-6ccf4cc873e5

zachdaniel commented on documentation/tutorials/remember-me.md:
```diff
@@ -0,0 +1,247 @@
+
+    # 2 of 3: Add the preparation. See the preparation docs if using an argument
+    # or strategy name other than remember_me
+    prepare AshAuthentication.Strategy.RememberMe.MaybeGenerateTokenPreparation
```
I assume this would need to be told what the new argument name is somehow if they don't use `remember_me`? Should we show that here?

rgraff replied:
Updated code comments

zachdaniel commented:
One other question: do we need to take this into account for log out everywhere? I think we might?

maennchen reviewed (changes requested):
Also add a test to verify that `AshAuthentication.AddOn.LogOutEverywhere` also invalidates remember me tokens.

rgraff commented on lib/ash_authentication/strategies/remember_me/cookie.ex:
```diff
@@ -0,0 +1,21 @@
+  Helper functions for working with remember me cookies.
+  """
+
+  @cookie_name_prefix "ash_auth:remember_me"
```
@zachdaniel I think I should get rid of the concept prefixing the cookie name. They can already set a custom name in the DSL.

rgraff commented:
I believe I'v addressed all the PR comments:

- [x] - Remove `ash_auth` as a prefix to the cookies as requested by @zachdaniel 
- [x] - Added test to verify that AshAuthentication.AddOn.LogOutEverywhere revokes remember me tokens @maennchen 
- [x] - Deleting remember me cookies also now revokes the associated cookie @maennchen 
- [x] - Fixes typos and reverted change to Dockerfile @jimsynz 
- [x] - Verified that the DSL for token lifetime is already applying validation @jimsynz

For clarity on token revocation:
* On sign out, remember me cookies/tokens in that _one_ connection are removed and revoked. If a user has remember me on other browsers, those cookies and tokens remain valid. This is the desired functionality.
* The "log out everywhere" feature which is typically called on password change, revokes all tokens including remember me.

Once this branch is merged, I'll put the final touches on the related PR for AshAuthenticationPhoenix.

I have verified that you can use RememberMe without AshAuthenticationPhoenix by following the docs in the `RememberMe.Plug.Helpers`

The final CI check is on the commit message. We'll need to a squash commit with a proper message.

rgraff commented on documentation/tutorials/remember-me.md:
```diff
@@ -0,0 +1,233 @@
+    # This will delete any remember_me tokens in the current connection's
+    # cookies and revoke them. Other remember_me tokens for the user are
+    # still valid.
+    |> delete_all_remember_me_cookies(:my_otp_app)
```
```suggestion
    #  Provided by `use AshAuthentication.Phoenix.Controller` or you can
    #  import AshAuthentication.Strategy.RememberMe.Plug.Helpers
    |> delete_all_remember_me_cookies(:my_otp_app)
```

rgraff commented on lib/ash_authentication/strategies/remember_me/token/helpers.ex:
```diff
@@ -0,0 +1,18 @@
+  def revoke_remember_me_token(token, otp_app, opts \\ []) do
+    with {:ok, resource} <- Jwt.token_to_resource(token, otp_app),
+         {:ok, token_resource} <- Info.authentication_tokens_token_resource(resource) do
+      # we want this to blow up if something goes wrong
+      :ok = TokenResource.Actions.revoke(token_resource, token, opts)
```
```suggestion
      TokenResource.Actions.revoke(token_resource, token, opts)
```

zachdaniel commented:
The commit message check is broken and I haven't had a chance to look into why 😢
